_posts/2023-06-20-introducing-iam-for-openfaas.md

```diff
@@ -0,0 +1,479 @@
+```
+
+The `policy` field contains a set of Policies to apply for this Role.
+
+The `condition` field can be used to limit permissions by matching fields in the jwt access token.
+
+Every condition must return true for the Role to be considered as a match.
+
+Valid conditions include: `StringEqual` or `StringLike`
+
+A user's email could also be fuzzy matched with a condition, for example:
+
+```yaml
+condition:
+    StringLike:
+      jwt:email: ["*@example.com"]
+```
+
+The principal filed is optional, however if it is given, both the principal and the condition must match. If the principal contains multiple items only one must match the token for it to be considered a match.
+
+In this example an access token will match the Role for 1 staff member with sub `aa544816-e4e9-4ea0-b4cf-dd70db159d2e` only if it was issued by `https://keycloak.example.com/realms/openfaas`.
+
+Create a policy:
+```yaml
+apiVersion: openfaas.com/v1
+kind: Policy
+metadata:
+  name: dev-rw
+  namespace: openfaas
+spec:
+  statement:
+  - sid: 1-rw-dev
+    action:
+    - Function:Read
+    - Function:Admin
+    - Secret:Read
+    effect: Allow
+    resource: ["dev:*"]
+```
+
+The Policy describes which permissions a user has, and on which resources.
+
+Permission can be scoped cluster wide or to specific namespaces. In this example the policy only applies for the `dev` namespace. To apply a policy cluster wide use `resource: ["*"]`.
+
+For an overview of the supported actions see: [permission](https://docs.openfaas.com/openfaas-pro/iam/overview/#permissions)
+
+### Setup the OpenFaaS dashboard
+
+The dashboard requires a couple of keys to be created and stored in Kubernetes secrets:
+
+- JWT signing key, used to sign and validate logged in sessions.
+- AES key, used to encrypt the OpenFaaS access token.
+
+Generate and store the JWT signing key:
+
+```bash
+# Generate a private key
+openssl ecparam -genkey -name prime256v1 -noout -out jwt_key
+
+# Then create a public key from the private key
+openssl ec -in jwt_key -pubout -out jwt_key.pub
+
+# Store both in a secret in the openfaas namespace
+kubectl -n openfaas \
+  create secret generic dashboard-jwt \
+  --from-file=key=./jwt_key \
+  --from-file=key.pub=./jwt_key.pub
+```
+
+Generate and store the AES encryption key:
+
+```bash
+# Generate a key
+openssl rand -hex 16 > aes_key
+
+# Store the key in a secret in the openfaas namespace
+kubectl -n openfaas \
+  create secret generic aes-key \
+  --from-file=aes_key=./aes_key
+```
+
+To execute the Authorization Code Flow the dashboard will need the client secret that you received when you configured your IdP.
+If your IdP does not need a client secret this step can be skipped.
+
+Store the Oauth client secret in a Kubernetes secret:
+
+```bash
+kubectl create secret generic \
+    -n openfaas \
+    auth0-client-secret \
+    --from-file client_secret=./client_secret
+```
+
+Add the following in your values.yaml file for the openfaas chart:
```

Comment:
Add the following to the values-iam.yaml file we created earlier.

A diff isn't necessary because the user isn't editing values.yaml from the master, but adding to a blank file that we created at the start.

Reply:
I wanted to make it cleat that we are adding the `dashboard` section and editing the `iam` section that we created earlier.